language/move-binary-format/src/layout.rs

```diff
@@ -0,0 +1,290 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    access::ModuleAccess,
+    file_format::{SignatureToken, StructDefinition, StructFieldInformation, StructHandleIndex},
+    CompiledModule,
+};
+use move_core_types::{
+    identifier::IdentStr,
+    language_storage::{ModuleId, StructTag, TypeTag},
+    resolver::ModuleResolver,
+    value::{MoveFieldLayout, MoveStructLayout, MoveTypeLayout},
+};
+use std::{cell::RefCell, collections::BTreeMap, fmt::Debug};
+
+/// A persistent storage that can fetch the bytecode for a given module id
+/// TODO: do we want to implement this in a way that allows clients to cache struct layouts?
+pub trait GetModule {
+    type Error: Debug;
+
+    fn get_module_by_id(&self, id: &ModuleId) -> Result<Option<CompiledModule>, Self::Error>;
+}
+
+/// Simple in-memory module cache
+pub struct ModuleCache<R> {
+    cache: RefCell<BTreeMap<ModuleId, CompiledModule>>,
+    resolver: R,
+}
+
+impl<R: ModuleResolver> ModuleCache<R> {
+    pub fn new(resolver: R) -> Self {
+        ModuleCache {
+            cache: RefCell::new(BTreeMap::new()),
+            resolver,
+        }
+    }
+
+    pub fn add(&self, id: ModuleId, m: CompiledModule) {
+        self.cache.borrow_mut().insert(id, m);
+    }
+}
+
+impl<R: ModuleResolver> GetModule for ModuleCache<R> {
+    type Error = R::Error;
+
+    fn get_module_by_id(&self, id: &ModuleId) -> Result<Option<CompiledModule>, Self::Error> {
+        Ok(Some(
+            self.cache
+                .borrow_mut()
+                .entry(id.clone())
+                .or_insert_with(|| {
+                    let module_bytes = self
+                        .resolver
+                        .get_module(id)
+                        .unwrap()
+                        .unwrap_or_else(|| panic!("Failure resolving module {:?}", id));
+                    CompiledModule::deserialize(&module_bytes)
+                        .unwrap_or_else(|_| panic!("Failure deserializing module {:?}", id))
+                })
+                .clone(),
+        ))
+    }
+}
+
+pub enum TypeLayoutBuilder {}
+pub enum StructLayoutBuilder {}
+
+#[derive(Copy, Clone, Debug)]
+enum LayoutType {
+    WithFields,
+    Runtime,
+}
+
+impl TypeLayoutBuilder {
+    /// Construct `TypeLayout` with fields from `t`.
+    /// Panics if `resolver` cannot resolve a module whose types are referenced directly or
+    /// transitively by `t`.
+    pub fn build_with_fields(t: &TypeTag, resolver: &impl GetModule) -> MoveTypeLayout {
```

Comment:
I think it would be better to return error to the caller instead of panics as it might be hard to add a panic handler for the caller.